docs/Validation-and-Serialization.md

```diff
@@ -355,24 +355,133 @@ fastify.schemaCompiler = function (schema) { return ajv.compile(schema) })
 ```
 _**Note:** If you use a custom instance of any validator (even Ajv), you have to add schemas to the validator instead of fastify, since fastify's default validator is no longer used, and fastify's `addSchema` method has no idea what validator you are using._
 
-But maybe you want to change the validation library. Perhaps you like `Joi`. In this case, you can use it to validate the url parameters, body, and query string!
+<a name="using-other-validation-libraries"></a>
+#### Using other validation libraries
+
+The `schemaCompiler` function makes it easy to substitute `ajv` with almost any Javascript validation library (`Joi`, `yup`, ...).
+
+However, in order to make your chosen validation engine play well with Fastify's request/response pipeline, the function returned by your `schemaCompiler` function should return an object with either :
+
+* in case of validation failure: an `error` property, filled with an instance of `Error` or a string that describes the validation error
+* in case of validation success: an `value` property, filled with the coerced value that passed the validation
+
+The examples below are therefore equivalent:
 
 ```js
 const Joi = require('joi')
 
+// Validation options to match ajv's baseline options used in Fastify
+const joiOptions = {
+  abortEarly: false, // return all errors
+  convert: true, // change data type of data to match type keyword
+  allowUnknown : false, // remove additional properties
+  noDefaults: false
+}
+
+const joiBodySchema = Joi.object().keys({
+  age: Joi.number().integer().required(),
+  sub: Joi.object().keys({
+    name: Joi.string().required()
+  }).required()
+})
+
+const joiSchemaCompiler = schema => data => {
+  // Joi `validate` function returns an object with an error property (if validation failed) and a value property (always present, coerced value if validation was successful)
+  let { error, value } = joiSchema.validate(data, joiOptions)
+  if (error) {
+    return { error }
+  } else {
+    return { value }
+  }
+}
+
+fastify.post('/the/url', {
+  schema: {
+    body: joiBodySchema
+  },
+  schemaCompiler: joiSchemaCompiler
+}, handler)
+```
+
+```js
+const yup = require('yup')
+
+// Validation options to match ajv's baseline options used in Fastify
+const yupOptions = {
+  strict: false,
+  abortEarly: false, // return all errors
+  stripUnknown: true, // remove additional properties
+  recursive: true
+}
+
+const yupBodySchema = yup.object({
+  age: yup.number().integer().required(),
+  sub: yup.object()shape({
```

Comment:
```suggestion
  sub: yup.object().shape({
```

Reply:
Oh! I now wonder if the entire section "Validation messages with other validation libraries" wouldn't be more adequate under the "Error handling" section, don't you think? With the heading changed to "Error handling with ??? external/alternate/other ???  validation libraries"

Reply:
Maybe in future, we could think a refactor in order to split this document into smaller pieces, but for now, I think that here is the right place in order to give the user a complete knowledge of validation usage 👍